File: VSCode/Codigo06_MODELO_ZN.py
import tclab
import numpy as np
import time
import matplotlib.pyplot as plt

# Conexão com o TCLab
a = tclab.TCLab()

# Tempo de execução em minutos
run_time = 20

# Número de ciclos (1 ciclo por segundo)
loops = int(60.0 * run_time)
tm = np.zeros(loops)            # Vetor para armazenar o tempo
T1 = np.zeros(loops)            # Temperatura real
T1_ModeloZN = np.zeros(loops)    # Temperatura modelo ZN
Q1 = np.ones(loops) * 50        # Potência fixa em 50%

a.Q1(50)    # Aplicar potência ao hardware
Q = 50      # Aplicar potência aos modelos

# Gráfico
plt.figure(figsize=(12, 8))
plt.ion()
plt.show()

# Loop principal
start_time = time.time()
prev_time = start_time
dt = 1.0  # intervalo de tempo em segundos

try:
    for i in range(loops):
        # Tempo de espera
        sleep_time = max(0.01, dt - (time.time() - prev_time))
        time.sleep(sleep_time)

        # Atualizar tempo
        t = time.time()
        prev_time = t
        tm[i] = t - start_time

        # Ler temperatura real do TCLab
        T1[i] = a.T1

        # --- Modelo ZN ---
        if i <= 0:
            T1_ModeloZN[i] = T1[0]  # condição inicial
        else:
            T1_ModeloZN[i] = T1_ModeloZN[i-1] + 0.00272 * np.exp(-tm[i - 15] / 217) * Q * np.heaviside(i - 15, 0)

        # Atualizar gráfico
        plt.clf()
        ax1 = plt.subplot(2, 1, 1)
        ax1.grid()
        plt.plot(tm[:i+1], T1[:i+1], 'r-', label='Temperatura Real', linewidth=2)
        plt.plot(tm[:i+1], T1_ModeloZN[:i+1], 'g--', label='Modelo Ziegler_Nichols', linewidth=2)
        plt.ylabel('Temperatura (°C)', fontsize=14)
        plt.legend(loc='best', fontsize=12)

        ax2 = plt.subplot(2, 1, 2)
        ax2.grid()
        plt.plot(tm[:i+1], Q1[:i+1], 'r-', label='Potência (%)', linewidth=2)
        plt.ylabel('Potência (%)', fontsize=14)
        plt.xlabel('Tempo (s)', fontsize=14)
        plt.legend(loc='best', fontsize=12)

        plt.tight_layout()
        plt.draw()
        plt.pause(0.05)

    # --- Cálculo do Erro e Salvamento ---
    erro_medio_ZN = np.mean(np.abs(T1 - T1_ModeloZN))
    print(f'Erro Médio Absoluto entre Temperatura Real e Modelo Ziegler_Nichols: {erro_medio_ZN:.2f} °C')

    # Salvar gráfico em PNG
    plt.savefig('Grafico_TCLab_Modelo_ZN.png', dpi=300, bbox_inches='tight')

    # Salvar dados em .txt
    np.savetxt('Dados_simulacao_06_ZN.txt', np.column_stack((tm, T1, T1_ModeloZN, Q1)),
               header='Tempo (s)\tTemperatura Real (°C)\tModelo ZN (°C)\tPotência (%)', fmt='%.2f', delimiter='\t')

except KeyboardInterrupt:
    print('Interrompido pelo usuário.')

except Exception as e:
    print(f'Erro: {e}')
    raise

finally:
    a.Q1(0)    # Desliga o aquecedor
    a.LED(0)   # Apaga LED
    a.close()
    print('Simulação concluída.')

File: MATLAB/Codigo06_modeloZN.m
clear; clc;

%% Inicialização do hardware
run('tclab.m');  % Arquivo com as funções T1C(), h1(), led()
% Garante o desligamento seguro do TCLab mesmo em caso de erro ou interrupção (Ctrl+C)
cleanupObj = onCleanup(@() desligarTCLab());

%% Parâmetros de Simulação
run_time = 20;                 % Tempo total da simulação (min)
loops = 60 * run_time;         % Número de ciclos (1 ciclo por segundo)
dt = 1.0;                      % Intervalo de tempo em segundos

% Inicialização das variáveis
tm = zeros(loops,1);           % Vetor para armazenar o tempo (s)
T1 = zeros(loops,1);           % Temperatura real (°C)
T1_ModeloZN = zeros(loops,1);   % Temperatura do modelo Ziegler-Nichols (°C)
Q1 = ones(loops,1) * 50;       % Potência fixa em 50%
% (u permanece não utilizado neste exemplo)
  
% Aplicar potência inicial ao hardware
h1(Q1(1));
Q = 50;  % Potência aplicada aos modelos

%% Criação do Gráfico
figure('Position', [100 100 800 600]);
drawnow;

%% Loop principal de aquisição e simulação
start_time = tic;
prev_time = toc(start_time);

try
    for i = 1:loops
        % Tempo de espera para que o intervalo seja de dt segundos
        sleep_time = max(0.01, dt - (toc(start_time) - prev_time));
        pause(sleep_time);
        
        % Atualizar tempo
        t = toc(start_time);
        prev_time = t;
        tm(i) = t;
        
        % Ler temperatura real do TCLab
        T1(i) = T1C();
        
        % --- Modelo Ziegler-Nichols (ZN) ---
        if i == 1
            % Condição inicial
            T1_ModeloZN(i) = T1(1);
        elseif i < 16
            % Para i < 16 (ou seja, enquanto o atraso não ocorre), manter valor anterior
            T1_ModeloZN(i) = T1_ModeloZN(i-1);
        else
            % A atualização ocorre somente após 15 ciclos; como i-15 > 0, heaviside(i-15) retorna 1
            T1_ModeloZN(i) = T1_ModeloZN(i-1) + 0.00272 * exp(-tm(i-15) / 217) * Q * heaviside(i-15);
        end
        
        % --- Atualizar Gráfico ---
        clf;
        subplot(2,1,1);
        plot(tm(1:i), T1(1:i), 'r-', 'LineWidth', 2); hold on;
        plot(tm(1:i), T1_ModeloZN(1:i), 'g--', 'LineWidth', 2);
        ylabel('Temperatura (°C)', 'FontSize', 14);
        legend('Temperatura Real', 'Modelo Ziegler\_Nichols', 'FontSize', 12);
        grid on;
        
        subplot(2,1,2);
        plot(tm(1:i), Q1(1:i), 'r-', 'LineWidth', 2);
        ylabel('Potência (%)', 'FontSize', 14);
        xlabel('Tempo (s)', 'FontSize', 14);
        legend('Potência (%)', 'FontSize', 12);
        grid on;
        
        drawnow;
    end
    
    % --- Cálculo do Erro e Salvamento ---
    erro_medio_ZN = mean(abs(T1 - T1_ModeloZN));
    fprintf('Erro Médio Absoluto entre Temperatura Real e Modelo Ziegler_Nichols: %.2f °C\n', erro_medio_ZN);
    
    % Salvar gráfico em PNG
    saveas(gcf, 'Grafico_TCLab_Modelo_ZN.png');
    
    % Salvar dados em arquivo TXT
    dados = [tm, T1, T1_ModeloZN, Q1];
    header = {'Tempo_s', 'Temperatura_Real_C', 'Modelo_ZN_C', 'Potencia_pct'};
    T_data = array2table(dados, 'VariableNames', header);
    writetable(T_data, 'Dados_simulacao_06_ZN.txt', 'Delimiter', '\t');
    
catch ME
    disp(['Erro: ' ME.message]);
    rethrow(ME);
end

disp('Simulação concluída.');

%% Função de desligamento seguro do TCLab
function desligarTCLab()
    try
        h1(0);    % Desliga o aquecedor
        led(0);   % Apaga o LED
        disp('Dispositivos TCLab desligados com segurança.');
    catch
        disp('Erro ao tentar desligar os dispositivos.');
    end
end
